schemas/pipeline_test.go

```diff
@@ -0,0 +1,229 @@
+package schemas
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestPipelineValidate(t *testing.T) {
+	t.Run("no errands", func(t *testing.T) {
+		p := Pipeline{Name: "no errands!"}
+		assert.EqualError(t, p.Validate(), "no errands specified in pipeline")
+	})
+
+	t.Run("errands with duplicate name", func(t *testing.T) {
+		p := Pipeline{
+			Name: "duplicate name",
+			Errands: []Errand{{
+				Name: "errand",
+			}, {
+				Name: "errand",
+			}},
+		}
+
+		assert.EqualError(t, p.Validate(), "duplicate name in errands list: errand")
+	})
+
+	t.Run("errand with self dependency", func(t *testing.T) {
+		p := Pipeline{
+			Name:    "self dependency",
+			Errands: []Errand{{Name: "single errand"}},
+			Dependencies: []PipelineDependency{{
+				Target:    "single errand",
+				DependsOn: "single errand",
+			}},
+		}
+
+		assert.EqualError(t, p.Validate(), "errand cannot depend on itself: single errand")
+	})
+
+	t.Run("dependency with invalid target", func(t *testing.T) {
+		p := Pipeline{
+			Name:    "self dependency",
+			Errands: []Errand{{Name: "single errand"}},
+			Dependencies: []PipelineDependency{{
+				Target:    "not a real errand",
+				DependsOn: "single errand",
+			}},
+		}
+
+		assert.EqualError(t, p.Validate(), "dependency references unknown errand name: not a real errand")
+	})
+
+	t.Run("dependency with invalid dependsOn", func(t *testing.T) {
+		p := Pipeline{
+			Name:    "self dependency",
+			Errands: []Errand{{Name: "single errand"}},
+			Dependencies: []PipelineDependency{{
+				Target:    "single errand",
+				DependsOn: "not a real dependency",
+			}},
+		}
+
+		assert.EqualError(t, p.Validate(), "dependency references unknown errand name: not a real dependency")
+	})
+
+	t.Run("simple dependency cycle | 2 errands", func(t *testing.T) {
+		/*
+			A <--> B
+		*/
+		p := Pipeline{
+			Name: "simple dependency cycle",
+			Errands: []Errand{
+				{Name: "A"},
+				{Name: "B"},
+			},
+			Dependencies: []PipelineDependency{
+				{Target: "A", DependsOn: "B"},
+				{Target: "B", DependsOn: "A"},
+			},
+		}
+
+		assert.EqualError(t, p.Validate(), "no independent errands found; a cycle must exist")
```

Comment:
the wording on this error is confusing. Should a cycle exist?

Reply:
no it's more like there are no independent errands, that means there has to be a dependency cycle in the pipeline. Updated the message to be more clear